### PR TITLE
feat(score): implements score agent and changes some api

### LINKED ISSUE #27: Implement ScoreAgent
Due Jul 22, 2022

### LINKED ISSUE #38: Change in score_agent public api
~A new issue to accommodate this change might be needed.~

_Originally posted by @EdwardHaoranLee in https://github.com/EdwardHaoranLee/WoodokuLearn/pull/36#discussion_r928144614_

### PR DESCRIPTION
**Pull request by Xiaorui-Huang**
Implemented score agent logic, updated some api and doctrings

### Motivation and context

Resolves: #27 

### Before submitting this PR, please make sure:
 - [ ] Your code builds cleanly without any errors

Still cannot test yet

### CODE REVIEW

File: Woodoku/src/entity/score_agent.py [hunk 1]
```diff
@@ -1,20 +1,44 @@
...
     __score: int
+    __streaks: int
 
     def __init__(self):
-        pass
+        self.__score = 0
+        self.__streaks = 0
 
-    def calculate_winning(self, group_info: Dict[str, int]) -> None:
+    def calculate_winning(self, blocks: int, groups: int) -> None:
```
Comment (EdwardHaoranLee) [inline Woodoku/src/entity/score_agent.py:14]: ~A new issue to accommodate this change might be needed.~

File: Woodoku/src/entity/score_agent.py [hunk 2]
```diff
@@ -1,20 +1,44 @@
...
+            self.__score += GROUP_POINTS
+            if self.__streaks:
+                self.__score += STREAK_POINTS * self.__streaks
+            self.__streaks += 1
+        else:
+            self.__streaks = 0
+
+        combos = groups - 1
+        if combos > 0:
+            self.__score += COMBO_POINTS * combos
```
Comment (EdwardHaoranLee) [inline Woodoku/src/entity/score_agent.py:41]: Test cases might be needed?

File: Woodoku/src/entity/woodoku_board.py [hunk 3]
```diff
@@ -159,28 +160,26 @@ def add_shape(self, shape: WoodokuShape, x: int, y: int) -> None:
...
-        self.__scoreAgent.calculate_winning(group_info)
+        groups, group_blocks = self.__find_groups()
+        self.__scoreAgent.calculate_winning(len(shape), groups)
         self.__representation.remove_blocks(group_blocks)
 
     def get_score(self) -> int:
         return self.__scoreAgent.get_score()
 
-    def __find_groups(self) -> Tuple[Dict[str, int], Set[Tuple[int, int]]]:
+    def __find_groups(self) -> Tuple[int, Set[Tuple[int, int]]]:
```
Comment (EdwardHaoranLee) [inline Woodoku/src/entity/woodoku_board.py:174]: @YANLING-H Changes in test cases might be required.

**Review by EdwardHaoranLee — DISMISSED** [review]

File: Woodoku/src/entity/score_agent.py (hunk 2 above)
Comment (Xiaorui-Huang) [reply Woodoku/src/entity/score_agent.py:41]: I've created a new issue for it #37

File: Woodoku/src/entity/woodoku_board.py (hunk 3 above)
Comment (Xiaorui-Huang) [reply Woodoku/src/entity/woodoku_board.py:174]: #32 @YANLING-H, there perhaps need some change

File: Woodoku/src/entity/score_agent.py (hunk 1 above)
Comment (Xiaorui-Huang) [reply Woodoku/src/entity/score_agent.py:14]: added to #38

**Review by EdwardHaoranLee — APPROVED** [review]

**Pull request merged**